### ISSUE TITLE
WandB-Logger drops all the logged values in training step for PyTorch Lightning

### ISSUE THREAD

**Issue by IsCoelacanth** [wandb/client#1507]
train-step:
```
def training_step(self, batch, batch_idx):
            x, y = batch
            logits, masks = self(x)
            loss = self.loss_func(logits, y)
            self.log(f"train-loss", loss, prog_bar=True)
            return {
                "loss": loss,
                "pred": torch.argmax(logits, -1),
                "labels": y,
            }
```
and on epoch end it logs the Precision and recall. 
after the validation step. I get a ` wandb: WARNING Step must only increase in log calls.  Step 379 < 380; dropping {train-loss: 0.494, 'train-precision': 0.9, 'train-recall': 1, 'epoch': 1}`

WandB version: 0.10.10
PyTorch-Lightning: 1.0.6

**Comment by borisdayma** [comment]
Hi @IsCoelacanth 
How are you logging the other parameters (precision and recall)?

**Comment by isaacrob** [comment]
I have this issue too! PyTorch Lightning resets step to 0 at the start of a new epoch, but wandb expects step to keep increasing and seems to ignore that we're on a new epoch. I don't know how to resolve this, it seems like a mismatch of expectations for the behavior of step.

**Comment by borisdayma** [comment]
@isaacrob step works as expected with me (increasing even when changing epoch). See [this colab](https://colab.research.google.com/drive/16d1uctGaw2y9KhGBlINNTsWpmlXdJwRW)

**Comment by isaacrob** [comment]
Interesting. Another issue I'm facing is that epochs is being logged as a chart, with per-step reporting and everything, which doesn't make any sense and doesn't seem to be happening in your notebook based off the screenshot. Here's a screenshot of mine: 
![Screenshot from 2020-11-19 17-17-30](https://user-images.githubusercontent.com/6684952/99743817-4e745680-2a8b-11eb-9ffa-73edb5c2ec43.png)

I wonder if this is related to it not handling epochs correctly?

**Comment by borisdayma** [comment]
Yes epoch is now a constant (it comes directly from Pytorch Lightning) though it should increase over time.
It looks like you only have one epoch, unless you try to reset the logger at each epoch.

See what I get from the colab I shared: [W&B run](https://wandb.ai/borisd13/MNIST/runs/1tjej3u4?workspace=user-borisd13)

**Comment by isaacrob** [comment]
Ok so that's not it. Like OP I'm getting this warning at the end of my first validation batch:
wandb: WARNING Step must only increase in log calls.  Step 199 < 201; dropping {'val_cls_loss': 1.3014671802520752, 'val_box_loss': 0.9644062519073486, 'val_masking_loss': 0.544899582862854, 'val_tot_loss': 2.8107728958129883, 'epoch': 0}.
and I don't see any metrics related to validation show up online, which appears to be because it's dropping all of them. It looks like then that step might be reset during validation call? When I reach the next epoch the same thing happens:
wandb: WARNING Step must only increase in log calls.  Step 399 < 401; dropping {'val_cls_loss': 1.0469279289245605, 'val_box_loss': 0.9169614315032959, 'val_masking_loss': 0.49405547976493835, 'val_tot_loss': 2.457944869995117, 'epoch': 1}.

**Comment by isaacrob** [comment]
Fixed it! I was calling `trainer.logger.experiment.log` in a callback and apparently this increases the step unless `commit` is set to `False`, in which case it gets group with the next logging message sent via PyTorch Lightning. Subtle bug! Everything's ok now and working as expected :) thanks for your help @borisdayma. Hope @IsCoelacanth was able to resolve their issue too!

**Comment by IsCoelacanth** [comment]
hey, @borisdayma there's a callback on validation completion which logs the metrics calculated using SKLearn metrics. same way @isaacrob mentioned. 
setting `commit = False` removes the warning but behaves the same way as mentioned by Isaac.

**Comment by borisdayma** [comment]
When you set `commit=False` or when you explicitly set a `step`, the values are aggregated and sent only when step increments (or at the end of the run) so you don't see them immediately in the dashboard.

Here is the logic:
* `wandb.log(dict, step=50)`
called automatically by `WandLogger` during training and validation
* **Option 1**: manual call to `wandb.log(my_dict)`
this will log at step 50, and increment to step 51 ; if you do multiple manual logs, step will increase, causing an issue if you try to log "back in time"
* **Option 2**: manual call to `wandb.log(my_dict, commit=False)`
this will log at step 50 and not increment
* **Option 3**: manual call to `wandb.log(my_dict, step=trainer.global_step)`
this will ensure your logging is associated to the right step

In practice, just setting `commit=False` should work well since you will probably do manual logging after a training step, which means all the metrics will be aggregated to last training step logged (which I guess is what you want).

Let me know what you think.
We could output a message suggesting to use `commit=False` when this situation occurs.

**Comment by vidalmaxime** [comment]
I am having the same issue (with the latest versions of pytorchlightning/wandb), although I do not have any manual logging/callbacks: `wandb: WARNING Step must only increase in log calls.  Step 116 < 117; dropping {'train_loss': 0.8560853600502014, 'prec': 0.3888888955116272, 'dist_acc': 0.013888888992369175,...}`
My training and validation steps are the following:

```
    def training_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self.forward(x)
        loss, metrics = self.criterion(y_hat, y)
        self.log_dict({"train_loss": loss, **metrics}, prog_bar=False)
        return loss

    def validation_step(self, batch, batch_idx):
        x, y = batch
        y_hat = self.forward(x)
        loss, metrics = self.criterion(y_hat, y)
        log = {"loss": loss, **metrics}
        self.log_dict(
            {f"val_{k}": v for k, v in log.items()}, prog_bar=False, on_epoch=True,
        )
        return {"y": y, "y_hat": y_hat}
```
If I stop logging in the validation step, then it's fine. Any ideas?

**Comment by borisdayma** [comment]
@vidalmaxime Does it work if you don't provide the `on_epoch` arg? I believe it should be set automatically for you (though in that case it would probably be `True`).
Also can you confirm you have latest version of Pytorch Lightning?

If those are the 2 only places where you log to `wandb`, it should have logged at the correct step.

I see you return values at `validation_step`, are you logging other values at `validation_epoch_end`?

**Comment by vidalmaxime** [comment]
Without the `on_epoch` arg, the issue persists.  PytorchLightning version is 1.1.1. Nothing logged at `validation_epoch_end`. 

I'm also using Hydra if that could be the cause ?

**Comment by borisdayma** [comment]
No it should still work unless you're logging other values through there.
You can also take a look at your dashboard and verify that all the metrics logged are the ones you expect.

I just run [an example] and it seemed to work all fine.

**Comment by vidalmaxime** [comment]
On wandb dashboard, the validation metrics look fine, but the training metrics stop after a number of steps. 
I get the warnings starting on epoch 3. Not sure...

**Comment by vidalmaxime** [comment]
Reverting to pytorch-lightning 1.0.8 fixes it :/

**Comment by borisdayma** [comment]
If it does not get resolved, feel free to share your code and maybe we can see where it comes from.

**Comment by LindaSt** [comment]
I have a similar issue, I'm logging in the following way:

```
def training_step(self, batch, batch_idx, **kwargs):
    loss, y_hat = self.forward(batch)
    # log step metric
    self.log('train_loss', loss)
    self.log('train_acc', self.train_acc(y_hat, batch.y))
    return loss

def validation_step(self, batch, batch_idx):
    loss, y_hat = self.forward(batch)
    self.log('val_loss', loss)
    self.log('val_acc', self.val_acc(y_hat, batch.y))
    return loss
```

The training loss and accuracy are logged step-wise to wanddb but nothing is logged from the validation phase. I get the following warning: `wandb: WARNING Step must only increase in log calls.` (full log below). I also get one warning regarding the training at the beginning, but nothing more later on. Adding `on_step=False, on_epoch=True` to the logging did not help either.

I've seen several issues concerning this here and on the pytroch-lightning GitHub page, but nothing seems to have been fixed so far. Has anybody found a solution for this yet?

```
Epoch 0:   1%|          | 1/100 [00:15<25:37, 15.54s/it, loss=0.588, v_num=1m1q]/Users/linda/miniconda2/envs/GraphNeuralNetworkExperiments/lib/python3.8/site-packages/pytorch_lightning/utilities/distributed.py:49: UserWarning: Trying to log at a previous step. Use `commit=False` when logging metrics manually.
  warnings.warn(*args, **kwargs)
wandb: WARNING Step must only increase in log calls.  Step 0 < 1; dropping {'train_loss': 0.5882745981216431, 'train_acc': 0.4375, 'epoch': 0}.
Epoch 0:  75%|███████▌  | 75/100 [00:18<00:06,  4.02it/s, loss=0.421, v_num=1m1q]
Validating: 0it [00:00, ?it/s]
Validating:   0%|          | 0/25 [00:00<?, ?it/s]
Epoch 0:  80%|████████  | 80/100 [00:23<00:05,  3.46it/s, loss=0.421, v_num=1m1q]
Epoch 0:  86%|████████▌ | 86/100 [00:23<00:03,  3.70it/s, loss=0.421, v_num=1m1q]
Epoch 0:  92%|█████████▏| 92/100 [00:23<00:02,  3.94it/s, loss=0.421, v_num=1m1q]
Epoch 0:  98%|█████████▊| 98/100 [00:23<00:00,  4.18it/s, loss=0.421, v_num=1m1q]
Epoch 0: 100%|██████████| 100/100 [00:24<00:00,  4.06it/s, loss=0.421, v_num=1m1q]
wandb: WARNING Step must only increase in log calls.  Step 74 < 75; dropping {'val_loss': 0.29613080620765686, 'val_acc': 0.8149999976158142, 'epoch': 0}.
Epoch 1:  75%|███████▌  | 75/100 [00:18<00:06,  4.04it/s, loss=0.265, v_num=1m1q]
Validating: 0it [00:00, ?it/s]
Validating:   0%|          | 0/25 [00:00<?, ?it/s]
Epoch 1:  79%|███████▉  | 79/100 [00:22<00:06,  3.44it/s, loss=0.265, v_num=1m1q]
Epoch 1:  85%|████████▌ | 85/100 [00:23<00:04,  3.69it/s, loss=0.265, v_num=1m1q]
Epoch 1:  91%|█████████ | 91/100 [00:23<00:02,  3.92it/s, loss=0.265, v_num=1m1q]
Epoch 1:  97%|█████████▋| 97/100 [00:23<00:00,  4.16it/s, loss=0.265, v_num=1m1q]
Epoch 1: 100%|██████████| 100/100 [00:24<00:00,  4.10it/s, loss=0.265, v_num=1m1q]
wandb: WARNING Step must only increase in log calls.  Step 149 < 150; dropping {'val_loss': 1.7754814624786377, 'val_acc': 0.9474999904632568, 'epoch': 1}.
Epoch 1: 100%|██████████| 100/100 [00:24<00:00,  4.08it/s, loss=0.265, v_num=1m1q]
Testing: 100%|██████████| 25/25 [00:07<00:00,  3.53it/s]
--------------------------------------------------------------------------------
DATALOADER:0 TEST RESULTS
{'test_acc': tensor(0.8425), 'test_loss': tensor(0.3146)}
--------------------------------------------------------------------------------
[{'test_loss': 0.3146221935749054, 'test_acc': 0.8424999713897705}]
```

**Comment by borisdayma** [comment]
For info https://github.com/PyTorchLightning/pytorch-lightning/pull/5351 just got merged so this problem will be fixed in next release of pytorch-lightning by using `sync_step=True`.

**Comment by borisdayma** [comment]
Working on a better fix #5931

**Comment by borisdayma** [comment]
This was fixed with https://github.com/PyTorchLightning/pytorch-lightning/pull/5931.
You can now just log at any time with `self.log('my_metric', my_value)` and you won't have any dropped value.
Just choose your x-axis appropriately in the UI (whether `global_step` or just the auto-incremented `step`).

**Comment by PhaelIshall** [comment]
@borisdayma I am using WANDB with pykeen (sdk) and every time I change the epoch it gives me the error of `Trying to change epoch, use wandb.config() with allow_val_change=True `

When I do that, it overwrites my results (for each epoch I am experimenting with several random seeds) and I only end up with one run saved per epoch. It was working fine a month ago or so, could I just go back to a previous version?

**Comment by borisdayma** [comment]
@PhaelIshall is there a run you can share?
How do you change the epoch?

**Comment by borisdayma** [comment]
@PhaelIshall if you're just trying to create a new W&B run within a same script, you should call `wandb.finish()` at the end of your run.

**Comment by ariG23498** [comment]
Hey @PhaelIshall 

We are closing this ticket as the original issue has been resolved. If you see any issue persisting, would you like to open a new issue to track the same? You could also comment below to reopen this issue. Thanks.
